# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/android/nsDragService.cpp
# Commit: 384b025f56d3
# Full Hash: 384b025f56d3f87e045ecede6b1a16b71c0b5d4f
# Author: David P <daparks@mozilla.com>
# Date: 2024-07-04 16:09:52
# Regressor Bug: 1893119
# File Overlap Count: 1
# Description:
#   Bug 1893119: Part 21 - Separate nsIDragService and nsIDragSession implementations r=gstoll,geckoview-reviewers,rkraesig,win-reviewers,m_kato
#   
#   Split the class inheritance trees for nsIDragService and nsIDragSession.
#   Remember that, before the start of this patch series, the inheritance
#   diagram was:
# ==============================================================================

diff -r beeda5ef3712 -r 384b025f56d3 widget/android/nsDragService.cpp
--- a/widget/android/nsDragService.cpp	Thu Jul 04 07:48:11 2024 +0000
+++ b/widget/android/nsDragService.cpp	Thu Jul 04 07:48:11 2024 +0000
@@ -20,9 +20,6 @@
 #include "nsViewManager.h"
 #include "nsWindow.h"
 
-NS_IMPL_ISUPPORTS_INHERITED0(nsDragSession, nsBaseDragService)
-NS_IMPL_ISUPPORTS_INHERITED0(nsDragService, nsDragSession)
-
 using namespace mozilla;
 using namespace mozilla::widget;
 
@@ -39,6 +36,11 @@
   return service.forget();
 }
 
+already_AddRefed<nsIDragSession> nsDragService::CreateDragSession() {
+  RefPtr<nsDragSession> session = new nsDragSession();
+  return session.forget();
+}
+
 static nsWindow* GetWindow(dom::Document* aDocument) {
   if (!aDocument) {
     return nullptr;
@@ -63,7 +65,7 @@
   return window.get();
 }
 
-nsresult nsDragService::InvokeDragSessionImpl(
+nsresult nsDragSession::InvokeDragSessionImpl(
     nsIWidget* aWidget, nsIArray* aTransferableArray,
     const Maybe<CSSIntRegion>& aRegion, uint32_t aActionType) {
   if (jni::GetAPIVersion() < 24) {
@@ -90,7 +92,6 @@
   if (nsWindow* window = GetWindow(mSourceDocument)) {
     mTransferable = transferable;
 
-    nsBaseDragService::StartDragSession(aWidget);
     OpenDragPopup();
 
     auto bitmap = CreateDragImage(mSourceNode, aRegion);
@@ -230,18 +231,12 @@
   mDataTransfer = nullptr;
 }
 
-// static
-void nsDragService::SetDropData(
+void nsDragSession::SetDropData(
     mozilla::java::GeckoDragAndDrop::DropData::Param aDropData) {
   MOZ_ASSERT(NS_IsMainThread());
 
-  RefPtr<nsDragService> dragService = nsDragService::GetInstance();
-  if (!dragService) {
-    return;
-  }
-
   if (!aDropData) {
-    dragService->SetData(nullptr);
+    SetData(nullptr);
     return;
   }
 
@@ -254,7 +249,7 @@
 
   if (!mime.EqualsLiteral("text/plain") && !mime.EqualsLiteral("text/html")) {
     // Not supported data.
-    dragService->SetData(nullptr);
+    SetData(nullptr);
     return;
   }
 
@@ -263,12 +258,12 @@
   nsPrimitiveHelpers::CreatePrimitiveForData(
       mime, buffer.get(), buffer.Length() * 2, getter_AddRefs(wrapper));
   if (!wrapper) {
-    dragService->SetData(nullptr);
+    SetData(nullptr);
     return;
   }
   nsCOMPtr<nsITransferable> transferable =
       do_CreateInstance("@mozilla.org/widget/transferable;1");
   transferable->Init(nullptr);
   transferable->SetTransferData(mime.get(), wrapper);
-  dragService->SetData(transferable);
+  SetData(transferable);
 }